refgenconf/helpers.py

```diff
@@ -69,3 +76,191 @@ def get_dir_digest(path, pm=None):
+    remove 'genome_digests' section from the config
+    replace all aliases in keys/asset names with genome digests
+
+    :param obj rgc: RefGenConfV03 obj
+    :param function(str, str) -> str get_json_url: how to build URL from
+            genome server URL base, genome, and asset
+    """
+
+    _LOGGER.info("Upgrade the v0.3 config file format to v0.4 format.")
+
+    missing_digest = []
+
+    # check if any genome lack of local fasta asset and not on server
+    for genome, genome_v in rgc[CFG_GENOMES_KEY].items():
+        digest = ""
+        # get genome digest from the server
+        cnt = 0
+        servers = rgc[CFG_SERVERS_KEY]
+        for server in servers:
+            cnt += 1
+            url_alias = get_json_url(server=server).format(alias=genome)
+            try:
+                digest = download_json(url_alias)
+                _LOGGER.info(
+                    f"Retrieve {genome} digest from the server.")
+            except DownloadJsonError:
+                if cnt == len(servers):
+                    try:
+                        tag = rgc.get_default_tag(genome, "fasta")
+                        asset_path = rgc.seek(genome, "fasta", tag, "fasta")
+                        ssc = SeqColClient({})
+                        digest, _ = ssc.load_fasta(asset_path)
+                        _LOGGER.info(
+                            f"Generate {genome} digest from local fasta file.")
+                    except MissingAssetError:
+                        _LOGGER.info(
+                            f"Fail to retrieve the digest for {genome}.")
+                        continue
+                continue
+
+        if digest:
+            # create "aliases" section
+            genome_v[CFG_ALIASES_KEY] = [genome]
+            # convert seek keys, children/parent asset keys from aliases to
+            # genome digests
+            for asset, asset_v in genome_v[CFG_ASSETS_KEY].items():
+                for tag, tag_v in asset_v[CFG_ASSET_TAGS_KEY].items():
+                    for seek, seek_v in tag_v[CFG_SEEK_KEYS_KEY].items():
+                        tag_v[CFG_SEEK_KEYS_KEY][seek] = seek_v.replace(genome,
+                                                                        digest)
+
+                    if CFG_ASSET_CHILDREN_KEY in tag_v:
+                        for i in range(len(asset_v[CFG_ASSET_TAGS_KEY][tag][
+                                               CFG_ASSET_CHILDREN_KEY])):
+                            tag_v[CFG_ASSET_CHILDREN_KEY][i] = \
+                            tag_v[CFG_ASSET_CHILDREN_KEY][i].replace(genome, digest)
+
+                    if CFG_ASSET_PARENTS_KEY in tag_v:
+                        for i in range(len(asset_v[CFG_ASSET_TAGS_KEY][tag][
+                                               CFG_ASSET_PARENTS_KEY])):
+                            tag_v[CFG_ASSET_PARENTS_KEY][i] = \
+                            tag_v[CFG_ASSET_PARENTS_KEY][i].replace(genome, digest)
```

Comment:
@xuebingjie1990 I've just implemented a recursive function in `refgenieserver` that replaces an arbitrary string in any object and I think this could be used in this code block. Can you check if this would make things easier for you here?

If it does, I'll add it in `refgenconf` and use in `refgenieserver`


```python
def replace_str_in_obj(object, x, y):
    """
    Replace strings in an object

    :param any object: object to replace strings in
    :param str x: string to replace
    :param str y: replacement
    :return any: object with strings replaced
    """
    _replace = partial(replace_str_in_obj, x=x, y=y)
    obj = copy(object)
    if isinstance(obj, dict):
        for k, v in obj.items():
            obj[k] = _replace(v)
    if isinstance(obj, list):
        obj = [_replace(i) for i in obj]
    if isinstance(object, str):
        obj = object.replace(x, y)
    return obj
```

Reply:
```
y=replace_str_dict(x, "511fb1178275e7d529560d53b949dba40815f195623bce8e", "rCRSd")
x
Out[32]: 
{'seek_keys': {'bowtie2_index': '511fb1178275e7d529560d53b949dba40815f195623bce8e'},
 'asset_parents': ['511fb1178275e7d529560d53b949dba40815f195623bce8e/fasta:default'],
 'asset_path': 'bowtie2_index',
 'asset_digest': '1262e30d4a87db9365d501de8559b3b4',
 'archive_digest': '61af83e05d4b409d82b781c04cd8aad7',
 'archive_size': '117.5KB',
 'asset_size': '8.1MB',
 'asset_children': []}
y
Out[33]: 
{'seek_keys': {'bowtie2_index': 'rCRSd'},
 'asset_parents': ['rCRSd/fasta:default'],
 'asset_path': 'bowtie2_index',
 'asset_digest': '1262e30d4a87db9365d501de8559b3b4',
 'archive_digest': '61af83e05d4b409d82b781c04cd8aad7',
 'archive_size': '117.5KB',
 'asset_size': '8.1MB',
 'asset_children': []}
```

Reply:
yes, can you commit that?

Reply:
done

Reply:
did you test the function with `replace_str_in_obj`? Is it working properly? Do we keep it in this package?

Reply:
yes it worked